ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/computation/DeprecatedStatementsChecker.java

```diff
@@ -0,0 +1,122 @@
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.rest.server.computation;
+
+import io.confluent.ksql.analyzer.Analysis;
+import io.confluent.ksql.engine.rewrite.DataSourceExtractor;
+import io.confluent.ksql.metastore.MetaStore;
+import io.confluent.ksql.metastore.model.DataSource;
+import io.confluent.ksql.parser.tree.AstNode;
+import io.confluent.ksql.parser.tree.CreateAsSelect;
+import io.confluent.ksql.parser.tree.Join;
+import io.confluent.ksql.parser.tree.JoinedSource;
+import io.confluent.ksql.parser.tree.Query;
+import io.confluent.ksql.parser.tree.Relation;
+import io.confluent.ksql.parser.tree.Statement;
+import io.confluent.ksql.parser.tree.WithinExpression;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+
+public class DeprecatedStatementsChecker {
+  public enum Deprecations {
+    DEPRECATED_STREAM_STREAM_OUTER_JOIN_WITH_NO_GRACE(
+        "DEPRECATION NOTICE: Left/Outer stream-stream joins statements without a GRACE PERIOD "
+            + "will not be accepted in a future ksqlDB version.\n"
+            + "Please use the GRACE PERIOD clause as specified in "
+            + "https://docs.ksqldb.io/en/latest/developer-guide/ksqldb-reference/"
+            + "select-push-query/"
+    );
+
+    private final String noticeMessage;
+
+    Deprecations(final String noticeMessage) {
+      this.noticeMessage = noticeMessage;
+    }
+
+    public String getNoticeMessage() {
+      return noticeMessage;
+    }
+  }
+
+  final MetaStore metaStore;
+
+  DeprecatedStatementsChecker(final MetaStore metaStore) {
+    this.metaStore = Objects.requireNonNull(metaStore, "metaStore");
+  }
+
+  public Optional<Deprecations> checkStatement(final Statement statement) {
+    if (isLeftOrOuterStreamStreamJoinWithoutGraceStatement(statement)) {
+      return Optional.of(Deprecations.DEPRECATED_STREAM_STREAM_OUTER_JOIN_WITH_NO_GRACE);
+    }
+
+    return Optional.empty();
+  }
+
+  private boolean isLeftOrOuterStreamStreamJoinWithoutGraceStatement(final Statement statement) {
+    if (statement instanceof CreateAsSelect) {
+      return isLeftOrOuterStreamStreamJoinWithoutGraceQuery(
+          ((CreateAsSelect) statement).getQuery());
+    } else if (statement instanceof Query) {
+      return isLeftOrOuterStreamStreamJoinWithoutGraceQuery((Query) statement);
+    }
+
+    return false;
+  }
+
+  private boolean isLeftOrOuterStreamStreamJoinWithoutGraceQuery(final Query query) {
+    if (query.getFrom() instanceof Join) {
+      final Join join = (Join) query.getFrom();
+
+      // check left joined source is a stream
+      if (!isStream(join.getLeft())) {
```

Comment:
Why can we exit early, if the left input is not a stream? -- Even if we don't support "busy" plans right now, it would be good to write generic logic IMHO to avoid future bugs.